connectors/sources/gitlab/datasource.py

```diff
@@ -0,0 +1,1040 @@
+        accessible_projects = set()
+
+        # Batch validate projects
+        for i in range(0, len(valid_project_paths), BATCH_SIZE):
+            batch = valid_project_paths[i : i + BATCH_SIZE]
+
+            try:
+                result = await self.gitlab_client._execute_graphql(
+                    VALIDATE_PROJECTS_QUERY, {"projectPaths": batch}
+                )
+
+                # Extract accessible project paths from response
+                projects_data = result.get("projects", {})
+                nodes = projects_data.get("nodes", [])
+
+                for node in nodes:
+                    full_path = node.get("fullPath")
+                    if full_path:
+                        accessible_projects.add(full_path)
+                        self._logger.debug(f"✓ Project '{full_path}' is accessible")
+
+            except Exception as e:
+                self._logger.warning(f"Failed to validate project batch {batch}: {e}")
+                # If batch query fails, we can't determine which specific projects are invalid
+                # so we skip this batch and continue
+
+        # Determine which projects are inaccessible
+        requested_projects = set(valid_project_paths)
+        invalid_projects = requested_projects - accessible_projects
+
+        for project in invalid_projects:
+            self._logger.warning(f"✗ Project '{project}' is not accessible")
+
+        if invalid_projects:
+            msg = f"The following projects are not accessible: {', '.join(sorted(invalid_projects))}. Please check the project paths and ensure your token has access."
+            raise ConfigurableFieldValueError(msg)
+
+    async def ping(self) -> None:
+        """Test the connection to GitLab."""
+        try:
+            await self.gitlab_client.ping()
+            self._logger.info("Successfully connected to GitLab.")
+        except Exception:
+            self._logger.exception("Error while connecting to GitLab.")
+            raise
+
+    async def close(self) -> None:
+        """Close the GitLab client connection."""
+        await self.gitlab_client.close()
+
+    def _should_sync_project(self, project_path):
+        """Check if a project should be synced based on configuration.
+
+        Args:
+            project_path (str): Full path of the project (e.g., 'group/project')
+
+        Returns:
+            bool: True if project should be synced, False otherwise
+        """
+        # If no projects configured or wildcard, sync all
+        if not self.configured_projects or "*" in self.configured_projects:
+            return True
+
+        # Check if project is in the configured list
+        return project_path in self.configured_projects
+
+    def _extract_widget_description(self, work_item: GitLabWorkItem) -> str | None:
+        """Extract description from work item widgets."""
+        for widget in work_item.widgets:
+            if isinstance(widget, WorkItemWidgetDescription):
```

Comment:
Is there a way to do it without casting or type checking? E.g. using some sort of discriminator field, rather than looping over the whole array multiple times to extract small amount of info?